presto-tests/src/test/java/com/facebook/presto/memory/TestMemoryManager.java

```diff
@@ -62,6 +65,77 @@
 
     private final ExecutorService executor = newCachedThreadPool();
 
+    @Test(timeOut = 240_000, expectedExceptions = ExecutionException.class, expectedExceptionsMessageRegExp = ".*The cluster is out of memory, and your query was killed. Please try again in a few minutes.")
+    public void testOutOfMemoryKiller()
+            throws Exception
+    {
+        Map<String, String> properties = ImmutableMap.<String, String>builder()
+                .put("task.verbose-stats", "true")
+                .put("task.operator-pre-allocated-memory", "0B")
+                .put("query.kill-on-out-of-memory-delay", "5s")
+                .put("query.kill-on-out-of-memory", "true")
+                .build();
+
+        try (DistributedQueryRunner queryRunner = createQueryRunner(TINY_SESSION, properties)) {
+            // Reserve all the memory
+            QueryId fakeQueryId = new QueryId("fake");
+            for (TestingPrestoServer server : queryRunner.getServers()) {
+                for (MemoryPool pool : server.getLocalMemoryManager().getPools()) {
+                    assertTrue(pool.tryReserve(fakeQueryId, pool.getMaxBytes()));
+                }
+            }
+
+            List<Future<?>> queryFutures = new ArrayList<>();
+            for (int i = 0; i < 2; i++) {
+                queryFutures.add(executor.submit(() -> queryRunner.execute("SELECT COUNT(*), clerk FROM orders GROUP BY clerk")));
+            }
+
+            ClusterMemoryManager memoryManager = queryRunner.getCoordinator().getClusterMemoryManager();
+            ClusterMemoryPool reservedPool;
+            while ((reservedPool = memoryManager.getPools().get(RESERVED_POOL)) == null) {
+                MILLISECONDS.sleep(10);
+            }
+
+            ClusterMemoryPool generalPool = memoryManager.getPools().get(GENERAL_POOL);
+            assertNotNull(generalPool);
+
+            // Wait for the cluster to run out of memory
+            while (generalPool.getAssignedQueries() != 1 || reservedPool.getAssignedQueries() != 1 || generalPool.getBlockedNodes() == 0 || reservedPool.getBlockedNodes() == 0) {
+                MILLISECONDS.sleep(10);
+            }
+
+            // Both queries should still be running
```

Comment:
so this code assumes that queries will take much longer than 5s, and the test code is much faster than 5s?  If so, maybe add a comment, since this will make this difficult to debug